src/provider.c

```diff
@@ -1045,12 +1049,30 @@ static int p11prov_module_init(P11PROV_CTX *ctx)
         .pReserved = (void *)ctx->init_args,
     };
     CK_INFO ck_info = { 0 };
+    const char *env_module;
     int ret;
 
     if (ctx->status != P11PROV_UNINITIALIZED) {
         return 0;
     }
 
+    /* The environment variable has the highest precedence */
+    env_module = getenv("PKCS11_PROVIDER_MODULE");
+    if (env_module && *env_module) {
+        OPENSSL_free(ctx->module);
+        ctx->module = OPENSSL_strdup(env_module);
+    }
+
+    /* If the module is not specified in the configuration file, use the p11-kit proxy  */
+    if (ctx->module == NULL) {
+#ifdef DEFAULT_PKCS11_MODULE
+        ctx->module = OPENSSL_strdup(DEFAULT_PKCS11_MODULE);
+#else
+        P11PROV_raise(ctx, -ENOENT, "No PKCS#11 module specified.");
```

Comment:
P!!PROV_raise take a CK_RV error which is then converted to txt via p11prov_get_reason_strings() by internal openssl dispatching.
So a (negative) errno is not ok, you should probably use  CKR_GENERAL_ERROR ... unless we want to start definining vendor specific errors and add them to the translation table.

Reply:
Thanks. Overlooked this. Fixed with the current version.